// @sanity/code-input/lib/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ACE_EDITOR_PROPS = exports.ACE_SET_OPTIONS = exports.DEFAULT_THEME = exports.SUPPORTED_THEMES = exports.LANGUAGE_ALIASES = exports.SUPPORTED_LANGUAGES = void 0;
var SUPPORTED_LANGUAGES = [{
  title: 'Batch file',
  value: 'batchfile'
}, {
  title: 'CSS',
  value: 'css'
}, {
  title: 'SASS',
  value: 'sass'
}, {
  title: 'SCSS',
  value: 'scss'
}, {
  title: 'HTML',
  value: 'html'
}, {
  title: 'JavaScript',
  value: 'javascript'
}, {
  title: 'TypeScript',
  value: 'typescript'
}, {
  title: 'JSON',
  value: 'json'
}, {
  title: 'JSX',
  value: 'jsx'
}, {
  title: 'Markdown',
  value: 'markdown'
}, {
  title: 'PHP',
  value: 'php'
}, {
  title: 'Python',
  value: 'python'
}, {
  title: 'sh',
  value: 'sh'
}, {
  title: 'Plain text',
  value: 'text'
}, {
  title: 'GROQ',
  value: 'groq'
}];
exports.SUPPORTED_LANGUAGES = SUPPORTED_LANGUAGES;
var LANGUAGE_ALIASES = {
  js: 'javascript'
};
exports.LANGUAGE_ALIASES = LANGUAGE_ALIASES;
var SUPPORTED_THEMES = ['github', 'monokai', 'terminal', 'tomorrow'];
exports.SUPPORTED_THEMES = SUPPORTED_THEMES;
var DEFAULT_THEME = 'tomorrow';
exports.DEFAULT_THEME = DEFAULT_THEME;
var ACE_SET_OPTIONS = {
  useSoftTabs: true,
  navigateWithinSoftTabs: true
  /* note only supported by ace v1.2.7 or higher */

};
exports.ACE_SET_OPTIONS = ACE_SET_OPTIONS;
var ACE_EDITOR_PROPS = {
  $blockScrolling: true
};
exports.ACE_EDITOR_PROPS = ACE_EDITOR_PROPS;
