// dtsgenerator/src/jsonSchemaParser.js
"use strict";

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var __awaiter = undefined && undefined.__awaiter || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) {
            try {
                step(generator.next(value));
            } catch (e) {
                reject(e);
            }
        }
        function rejected(value) {
            try {
                step(generator.throw(value));
            } catch (e) {
                reject(e);
            }
        }
        function step(result) {
            result.done ? resolve(result.value) : new P(function (resolve) {
                resolve(result.value);
            }).then(fulfilled, rejected);
        }
        step((generator = generator.apply(thisArg, _arguments)).next());
    });
};
var Debug = require('debug');
var request = require('request');
var JsonPointer = require('./jsonPointer');
var schemaid_1 = require('./schemaid');
var typeDefinition_1 = require('./typeDefinition');
var writeProcessor_1 = require('./writeProcessor');
var debug = Debug('dtsgen');
var walkMaker = '<<type>>';

var JsonSchemaParser = function () {
    function JsonSchemaParser() {
        _classCallCheck(this, JsonSchemaParser);

        this.typeCache = new Map();
        this.schemaReference = new Map();
        this.referenceCache = new Map();
    }

    _createClass(JsonSchemaParser, [{
        key: 'generateDts',
        value: function generateDts(prefix, header) {
            return __awaiter(this, void 0, Promise, regeneratorRuntime.mark(function _callee() {
                var _this = this;

                var _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, typeId, _iteratorNormalCompletion2, _didIteratorError2, _iteratorError2, _iterator2, _step2, ref, _iteratorNormalCompletion3, _didIteratorError3, _iteratorError3, _iterator3, _step3, schema, _iteratorNormalCompletion4, _didIteratorError4, _iteratorError4, _iterator4, _step4, id, process, env;

                return regeneratorRuntime.wrap(function _callee$(_context) {
                    while (1) {
                        switch (_context.prev = _context.next) {
                            case 0:
                                debug('generate d.ts: prefix=[' + prefix + '].');
                                _context.next = 3;
                                return this.resolveReference();

                            case 3:
                                debug('TypeId list:');
                                _iteratorNormalCompletion = true;
                                _didIteratorError = false;
                                _iteratorError = undefined;
                                _context.prev = 7;
                                for (_iterator = this.typeCache.keys()[Symbol.iterator](); !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                                    typeId = _step.value;

                                    debug('  ' + typeId);
                                }
                                _context.next = 15;
                                break;

                            case 11:
                                _context.prev = 11;
                                _context.t0 = _context['catch'](7);
                                _didIteratorError = true;
                                _iteratorError = _context.t0;

                            case 15:
                                _context.prev = 15;
                                _context.prev = 16;

                                if (!_iteratorNormalCompletion && _iterator.return) {
                                    _iterator.return();
                                }

                            case 18:
                                _context.prev = 18;

                                if (!_didIteratorError) {
                                    _context.next = 21;
                                    break;
                                }

                                throw _iteratorError;

                            case 21:
                                return _context.finish(18);

                            case 22:
                                return _context.finish(15);

                            case 23:
                                debug('SchemaId list:');
                                _iteratorNormalCompletion2 = true;
                                _didIteratorError2 = false;
                                _iteratorError2 = undefined;
                                _context.prev = 27;
                                for (_iterator2 = this.schemaReference.keys()[Symbol.iterator](); !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
                                    ref = _step2.value;

                                    debug('  ' + ref);
                                }
                                _context.next = 35;
                                break;

                            case 31:
                                _context.prev = 31;
                                _context.t1 = _context['catch'](27);
                                _didIteratorError2 = true;
                                _iteratorError2 = _context.t1;

                            case 35:
                                _context.prev = 35;
                                _context.prev = 36;

                                if (!_iteratorNormalCompletion2 && _iterator2.return) {
                                    _iterator2.return();
                                }

                            case 38:
                                _context.prev = 38;

                                if (!_didIteratorError2) {
                                    _context.next = 41;
                                    break;
                                }

                                throw _iteratorError2;

                            case 41:
                                return _context.finish(38);

                            case 42:
                                return _context.finish(35);

                            case 43:
                                debug('Reference list:');
                                _iteratorNormalCompletion3 = true;
                                _didIteratorError3 = false;
                                _iteratorError3 = undefined;
                                _context.prev = 47;
                                _iterator3 = this.referenceCache.keys()[Symbol.iterator]();

                            case 49:
                                if (_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done) {
                                    _context.next = 74;
                                    break;
                                }

                                schema = _step3.value;

                                debug('  ' + schema.id);
                                _iteratorNormalCompletion4 = true;
                                _didIteratorError4 = false;
                                _iteratorError4 = undefined;
                                _context.prev = 55;
                                for (_iterator4 = this.referenceCache.get(schema).keys()[Symbol.iterator](); !(_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done); _iteratorNormalCompletion4 = true) {
                                    id = _step4.value;

                                    debug('    ' + id);
                                }
                                _context.next = 63;
                                break;

                            case 59:
                                _context.prev = 59;
                                _context.t2 = _context['catch'](55);
                                _didIteratorError4 = true;
                                _iteratorError4 = _context.t2;

                            case 63:
                                _context.prev = 63;
                                _context.prev = 64;

                                if (!_iteratorNormalCompletion4 && _iterator4.return) {
                                    _iterator4.return();
                                }

                            case 66:
                                _context.prev = 66;

                                if (!_didIteratorError4) {
                                    _context.next = 69;
                                    break;
                                }

                                throw _iteratorError4;

                            case 69:
                                return _context.finish(66);

                            case 70:
                                return _context.finish(63);

                            case 71:
                                _iteratorNormalCompletion3 = true;
                                _context.next = 49;
                                break;

                            case 74:
                                _context.next = 80;
                                break;

                            case 76:
                                _context.prev = 76;
                                _context.t3 = _context['catch'](47);
                                _didIteratorError3 = true;
                                _iteratorError3 = _context.t3;

                            case 80:
                                _context.prev = 80;
                                _context.prev = 81;

                                if (!_iteratorNormalCompletion3 && _iterator3.return) {
                                    _iterator3.return();
                                }

                            case 83:
                                _context.prev = 83;

                                if (!_didIteratorError3) {
                                    _context.next = 86;
                                    break;
                                }

                                throw _iteratorError3;

                            case 86:
                                return _context.finish(83);

                            case 87:
                                return _context.finish(80);

                            case 88:
                                process = new writeProcessor_1.WriteProcessor(function (baseSchema, ref) {
                                    debug('Search Reference: schemaId=' + (baseSchema ? baseSchema.id : null) + ', ref=' + ref);
                                    var map = _this.referenceCache.get(baseSchema);
                                    if (map == null) {
                                        return undefined;
                                    }
                                    var refId = new schemaid_1.SchemaId(ref);
                                    var result = map.get(refId.getAbsoluteId());
                                    if (result == null) {
                                        if (refId.isJsonPointerHash()) {
                                            var fileId = refId.getFileId();
                                            var schema = fileId ? _this.schemaReference.get(fileId).targetSchema : baseSchema;
                                            debug('  fileId=' + fileId + ', schemaId=' + schema.id + '.');
                                            return JsonPointer.get(schema, refId.getJsonPointerHash());
                                        }
                                    }
                                    return result;
                                }, prefix);
                                env = this.createHierarchicalMap(this.typeCache);

                                if (header) {
                                    process.outputLine(header);
                                }
                                this.walk(process, env);
                                return _context.abrupt('return', process.toDefinition());

                            case 93:
                            case 'end':
                                return _context.stop();
                        }
                    }
                }, _callee, this, [[7, 11, 15, 23], [16,, 18, 22], [27, 31, 35, 43], [36,, 38, 42], [47, 76, 80, 88], [55, 59, 63, 71], [64,, 66, 70], [81,, 83, 87]]);
            }));
        }
    }, {
        key: 'createHierarchicalMap',
        value: function createHierarchicalMap(types) {
            var map = {};
            if (types.size === 0) {
                throw new Error('There is no id in the input schema(s)');
            }
            var _iteratorNormalCompletion5 = true;
            var _didIteratorError5 = false;
            var _iteratorError5 = undefined;

            try {
                for (var _iterator5 = types.values()[Symbol.iterator](), _step5; !(_iteratorNormalCompletion5 = (_step5 = _iterator5.next()).done); _iteratorNormalCompletion5 = true) {
                    var type = _step5.value;

                    var names = type.schemaId.getTypeNames();
                    JsonPointer.set(map, names.concat(walkMaker), type);
                }
            } catch (err) {
                _didIteratorError5 = true;
                _iteratorError5 = err;
            } finally {
                try {
                    if (!_iteratorNormalCompletion5 && _iterator5.return) {
                        _iterator5.return();
                    }
                } finally {
                    if (_didIteratorError5) {
                        throw _iteratorError5;
                    }
                }
            }

            return map;
        }
    }, {
        key: 'walk',
        value: function walk(process, env) {
            var _this2 = this;

            var path = arguments.length <= 2 || arguments[2] === undefined ? [] : arguments[2];

            var keys = Object.keys(env).sort();
            keys.forEach(function (key) {
                var val = env[key];
                var type = val[walkMaker];
                if (type instanceof typeDefinition_1.TypeDefinition) {
                    debug('  walk doProcess: path=' + JSON.stringify(path) + ', schemaId=' + type.schemaId.getAbsoluteId());
                    type.doProcess(process);
                }
                delete val[walkMaker];
                if (Object.keys(val).length > 0) {
                    var nextPath = path.concat(key);
                    if (process.indentLevel === 0) {
                        process.output('declare ');
                    }
                    process.output('namespace ').outputType(key, true).outputLine(' {');
                    process.increaseIndent();
                    _this2.walk(process, val, nextPath);
                    process.decreaseIndent();
                    process.outputLine('}');
                }
            });
        }
    }, {
        key: 'resolveReference',
        value: function resolveReference() {
            return __awaiter(this, void 0, Promise, regeneratorRuntime.mark(function _callee2() {
                var error, _iteratorNormalCompletion6, _didIteratorError6, _iteratorError6, _iterator6, _step6, schema, map, _iteratorNormalCompletion7, _didIteratorError7, _iteratorError7, _iterator7, _step7, _step7$value, ref, type, refId, fileId, fetchedSchema, pointer, targetSchema, _type, target;

                return regeneratorRuntime.wrap(function _callee2$(_context2) {
                    while (1) {
                        switch (_context2.prev = _context2.next) {
                            case 0:
                                debug('resolve reference: reference schema count=' + this.referenceCache.size + '.');
                                error = [];
                                _iteratorNormalCompletion6 = true;
                                _didIteratorError6 = false;
                                _iteratorError6 = undefined;
                                _context2.prev = 5;
                                _iterator6 = this.referenceCache.keys()[Symbol.iterator]();

                            case 7:
                                if (_iteratorNormalCompletion6 = (_step6 = _iterator6.next()).done) {
                                    _context2.next = 73;
                                    break;
                                }

                                schema = _step6.value;
                                map = this.referenceCache.get(schema);
                                _iteratorNormalCompletion7 = true;
                                _didIteratorError7 = false;
                                _iteratorError7 = undefined;
                                _context2.prev = 13;
                                _iterator7 = map[Symbol.iterator]();

                            case 15:
                                if (_iteratorNormalCompletion7 = (_step7 = _iterator7.next()).done) {
                                    _context2.next = 56;
                                    break;
                                }

                                _step7$value = _slicedToArray(_step7.value, 2);
                                ref = _step7$value[0];
                                type = _step7$value[1];

                                if (!(type != null)) {
                                    _context2.next = 21;
                                    break;
                                }

                                return _context2.abrupt('continue', 53);

                            case 21:
                                refId = new schemaid_1.SchemaId(ref);
                                fileId = refId.getFileId();

                                if (!(fileId && !this.schemaReference.has(fileId))) {
                                    _context2.next = 39;
                                    break;
                                }

                                if (refId.isFetchable()) {
                                    _context2.next = 27;
                                    break;
                                }

                                error.push('$ref target is not found: ' + ref);
                                return _context2.abrupt('continue', 53);

                            case 27:
                                _context2.prev = 27;

                                debug('fetch remote schema: id=[' + fileId + '].');
                                _context2.next = 31;
                                return this.fetchRemoteSchema(fileId);

                            case 31:
                                fetchedSchema = _context2.sent;

                                this.parseSchema(fetchedSchema, fileId);
                                _context2.next = 39;
                                break;

                            case 35:
                                _context2.prev = 35;
                                _context2.t0 = _context2['catch'](27);

                                error.push('fail to fetch the $ref target: ' + ref + ', ' + _context2.t0);
                                return _context2.abrupt('continue', 53);

                            case 39:
                                debug('resolve reference: ref=[' + ref + ']');

                                if (!refId.isJsonPointerHash()) {
                                    _context2.next = 48;
                                    break;
                                }

                                pointer = refId.getJsonPointerHash();
                                targetSchema = fileId ? this.schemaReference.get(fileId).rootSchema : schema;
                                _type = new typeDefinition_1.TypeDefinition(targetSchema, pointer, refId);

                                map.set(ref, _type);
                                this.addType(_type);
                                _context2.next = 53;
                                break;

                            case 48:
                                target = this.typeCache.get(ref);

                                if (!(target == null)) {
                                    _context2.next = 52;
                                    break;
                                }

                                error.push('$ref target is not found: ' + ref);
                                return _context2.abrupt('continue', 53);

                            case 52:
                                map.set(ref, target);

                            case 53:
                                _iteratorNormalCompletion7 = true;
                                _context2.next = 15;
                                break;

                            case 56:
                                _context2.next = 62;
                                break;

                            case 58:
                                _context2.prev = 58;
                                _context2.t1 = _context2['catch'](13);
                                _didIteratorError7 = true;
                                _iteratorError7 = _context2.t1;

                            case 62:
                                _context2.prev = 62;
                                _context2.prev = 63;

                                if (!_iteratorNormalCompletion7 && _iterator7.return) {
                                    _iterator7.return();
                                }

                            case 65:
                                _context2.prev = 65;

                                if (!_didIteratorError7) {
                                    _context2.next = 68;
                                    break;
                                }

                                throw _iteratorError7;

                            case 68:
                                return _context2.finish(65);

                            case 69:
                                return _context2.finish(62);

                            case 70:
                                _iteratorNormalCompletion6 = true;
                                _context2.next = 7;
                                break;

                            case 73:
                                _context2.next = 79;
                                break;

                            case 75:
                                _context2.prev = 75;
                                _context2.t2 = _context2['catch'](5);
                                _didIteratorError6 = true;
                                _iteratorError6 = _context2.t2;

                            case 79:
                                _context2.prev = 79;
                                _context2.prev = 80;

                                if (!_iteratorNormalCompletion6 && _iterator6.return) {
                                    _iterator6.return();
                                }

                            case 82:
                                _context2.prev = 82;

                                if (!_didIteratorError6) {
                                    _context2.next = 85;
                                    break;
                                }

                                throw _iteratorError6;

                            case 85:
                                return _context2.finish(82);

                            case 86:
                                return _context2.finish(79);

                            case 87:
                                if (!(error.length > 0)) {
                                    _context2.next = 89;
                                    break;
                                }

                                throw new Error(error.join('\n'));

                            case 89:
                                return _context2.abrupt('return', true);

                            case 90:
                            case 'end':
                                return _context2.stop();
                        }
                    }
                }, _callee2, this, [[5, 75, 79, 87], [13, 58, 62, 70], [27, 35], [63,, 65, 69], [80,, 82, 86]]);
            }));
        }
    }, {
        key: 'fetchRemoteSchema',
        value: function fetchRemoteSchema(fileId) {
            return new Promise(function (resolve, reject) {
                request.get(fileId, function (err, response, body) {
                    if (err) {
                        return reject(err);
                    } else if (response.statusCode !== 200) {
                        return reject(body);
                    } else {
                        try {
                            resolve(JSON.parse(body));
                        } catch (e) {
                            reject(e);
                        }
                    }
                });
            });
        }
    }, {
        key: 'parseSchema',
        value: function parseSchema(schema, url) {
            var _this3 = this;

            if (typeof schema === 'string') {
                schema = JSON.parse(schema);
            }
            debug('parse schema: schemaId=[' + schema.id + '], url=[' + url + '].');
            if (schema.id == null) {
                schema.id = url;
            }
            var walk = function walk(s, paths) {
                function walkArray(array, pathArray) {
                    array.forEach(function (item, index) {
                        walk(item, pathArray.concat(index.toString()));
                    });
                }
                function walkObject(obj, pathObject) {
                    var isDefinitions = arguments.length <= 2 || arguments[2] === undefined ? false : arguments[2];

                    Object.keys(obj).forEach(function (key) {
                        var sub = obj[key];
                        if (sub != null) {
                            if (isDefinitions && sub.id == null) {
                                sub.id = '#/definitions/' + key;
                            }
                            walk(sub, pathObject.concat(key));
                        }
                    });
                }
                if (s == null || (typeof s === 'undefined' ? 'undefined' : _typeof(s)) !== 'object') {
                    return;
                }
                var allOf = s.allOf;
                if (allOf != null) {
                    walkArray(allOf, paths.concat('allOf'));
                }
                var anyOf = s.anyOf;
                if (anyOf != null) {
                    walkArray(anyOf, paths.concat('anyOf'));
                }
                var oneOf = s.oneOf;
                if (oneOf != null) {
                    walkArray(oneOf, paths.concat('oneOf'));
                }
                var items = s.items;
                if (items != null) {
                    if (Array.isArray(items)) {
                        walkArray(items, paths.concat('items'));
                    } else {
                        walk(items, paths.concat('items'));
                    }
                }
                var additionalItems = s.additionalItems;
                if (additionalItems != null && typeof additionalItems !== 'boolean') {
                    walk(additionalItems, paths.concat('additionalItems'));
                }
                var definitions = s.definitions;
                if (definitions != null) {
                    walkObject(s.definitions, paths.concat('definitions'), true);
                }
                var properties = s.properties;
                if (properties != null) {
                    walkObject(s.properties, paths.concat('properties'));
                }
                var patternProperties = s.patternProperties;
                if (patternProperties != null) {
                    walkObject(s.patternProperties, paths.concat('patternProperties'));
                }
                var additionalProperties = s.additionalProperties;
                if (additionalProperties != null && typeof additionalProperties !== 'boolean') {
                    walk(additionalProperties, paths.concat('additionalProperties'));
                }
                if (typeof s.id === 'string') {
                    var type = new typeDefinition_1.TypeDefinition(schema, paths);
                    s.id = type.schemaId.getAbsoluteId();
                    _this3.addType(type);
                }
                if (typeof s.$ref === 'string') {
                    s.$ref = _this3.addReference(schema, s.$ref);
                }
            };
            walk(schema, []);
        }
    }, {
        key: 'addType',
        value: function addType(g) {
            var id = g.schemaId;
            if (id) {
                this.typeCache.set(id.getAbsoluteId(), g);
                debug('add type: id=' + id.getAbsoluteId());
                var fileId = id.getFileId();
                if (!this.schemaReference.has(fileId)) {
                    this.schemaReference.set(fileId, g);
                }
            }
        }
    }, {
        key: 'addReference',
        value: function addReference(schema, ref) {
            var map = this.referenceCache.get(schema);
            if (map == null) {
                map = new Map();
                this.referenceCache.set(schema, map);
            }
            var refId = new schemaid_1.SchemaId(ref, [schema.id]);
            map.set(refId.getAbsoluteId(), null);
            return refId.getAbsoluteId();
        }
    }, {
        key: 'clear',
        value: function clear() {
            debug('clear data cache.');
            this.typeCache.clear();
            this.schemaReference.clear();
            this.referenceCache.clear();
        }
    }]);

    return JsonSchemaParser;
}();

exports.JsonSchemaParser = JsonSchemaParser;
