// dtsgenerator/src/writeProcessor.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var WriteProcessor = function () {
    function WriteProcessor(refResolver) {
        var typePrefix = arguments.length <= 1 || arguments[1] === undefined ? '' : arguments[1];

        _classCallCheck(this, WriteProcessor);

        this.refResolver = refResolver;
        this.typePrefix = typePrefix;
        this.indentChar = ' ';
        this.indentStep = 4;
        this.indent = 0;
        this.results = '';
        this.alreadlyIndentThisLine = false;
        this.referenceStack = [];
    }

    _createClass(WriteProcessor, [{
        key: 'pushReference',
        value: function pushReference(referenceName) {
            return this.referenceStack.push(referenceName);
        }
    }, {
        key: 'popReference',
        value: function popReference() {
            return this.referenceStack.pop();
        }
    }, {
        key: 'checkCircularReference',
        value: function checkCircularReference(referenceName) {
            return this.referenceStack.indexOf(referenceName) < 0;
        }
    }, {
        key: 'output',
        value: function output(str) {
            this.doIndent();
            this.results += str;
            return this;
        }
    }, {
        key: 'outputType',
        value: function outputType(type) {
            var primitive = arguments.length <= 1 || arguments[1] === undefined ? false : arguments[1];

            if (this.typePrefix && !primitive) {
                this.output(this.typePrefix);
            }
            type = type.replace(/[^0-9A-Za-z_$]/g, '_');
            if (/^\d/.test(type)) {
                type = '$' + type;
            }
            this.output(type);
            return this;
        }
    }, {
        key: 'outputKey',
        value: function outputKey(name) {
            var optional = arguments.length <= 1 || arguments[1] === undefined ? false : arguments[1];

            if (/[^0-9A-Za-z_$]/.test(name) || /^\d/.test(name)) {
                this.output('\"').output(name).output('\"');
            } else {
                this.output(name);
            }
            if (optional) {
                this.output('?');
            }
            return this;
        }
    }, {
        key: 'outputLine',
        value: function outputLine(str) {
            this.doIndent();
            if (str) {
                this.output(str);
            }
            this.output('\n');
            this.alreadlyIndentThisLine = false;
            return this;
        }
    }, {
        key: 'outputJSDoc',
        value: function outputJSDoc(spec) {
            var _this = this;

            var description = spec.description;
            var example = spec.example;

            if (!description && !example) {
                return this;
            }
            this.outputLine('/**');
            if (description) {
                description.toString().split('\n').forEach(function (line) {
                    _this.output(' * ').outputLine(line);
                });
            }
            if (example) {
                var split = example.toString().split('\n');
                if (split.length === 1) {
                    this.outputLine(' * example: ' + example);
                } else {
                    this.outputLine(' * example:');
                    split.forEach(function (line) {
                        _this.output(' *   ').outputLine(line);
                    });
                }
            }
            this.outputLine(' */');
            return this;
        }
    }, {
        key: 'doIndent',
        value: function doIndent() {
            if (!this.alreadlyIndentThisLine) {
                var indent = this.getIndent();
                this.results += indent;
                this.alreadlyIndentThisLine = true;
            }
            return this;
        }
    }, {
        key: 'increaseIndent',
        value: function increaseIndent() {
            this.indent++;
            return this;
        }
    }, {
        key: 'decreaseIndent',
        value: function decreaseIndent() {
            this.indent--;
            return this;
        }
    }, {
        key: 'getIndent',
        value: function getIndent() {
            return this.repeatString(this.indent * this.indentStep, this.indentChar);
        }
    }, {
        key: 'repeatString',
        value: function repeatString(n, s) {
            var result = '';
            for (var i = 0; i < n; i++) {
                result += s;
            }
            return result;
        }
    }, {
        key: 'toDefinition',
        value: function toDefinition() {
            return this.results;
        }
    }, {
        key: 'referenceResolve',
        get: function get() {
            return this.refResolver;
        }
    }, {
        key: 'indentLevel',
        get: function get() {
            return this.indent;
        }
    }]);

    return WriteProcessor;
}();

exports.WriteProcessor = WriteProcessor;
