// react-responsive-modal/lib/css.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = ReactResponsiveModalCss;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _modal = require('./modal');

var _modal2 = _interopRequireDefault(_modal);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function ReactResponsiveModalCss(_ref) {
  var props = _objectWithoutProperties(_ref, []);

  return _react2.default.createElement(_modal2.default, _extends({
    classes: {
      overlay: 'react-responsive-modal-overlay',
      overlayLittle: 'react-responsive-modal-overlay-little',
      modal: 'react-responsive-modal-modal',
      closeIcon: 'react-responsive-modal-close-icon',
      transitionEnter: 'react-responsive-modal-transition-enter',
      transitionEnterActive: 'react-responsive-modal-transition-enter-active',
      transitionExit: 'react-responsive-modal-transition-exit',
      transitionExitActive: 'react-responsive-modal-transition-exit-active'
    }
  }, props));
}
