// @glimmer/util/dist/commonjs/es2017/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _platformUtils = require('./lib/platform-utils');

Object.keys(_platformUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _platformUtils[key];
    }
  });
});

var _assert = require('./lib/assert');

Object.defineProperty(exports, 'assert', {
  enumerable: true,
  get: function () {
    return _interopRequireDefault(_assert).default;
  }
});

var _objectUtils = require('./lib/object-utils');

Object.defineProperty(exports, 'assign', {
  enumerable: true,
  get: function () {
    return _objectUtils.assign;
  }
});
Object.defineProperty(exports, 'fillNulls', {
  enumerable: true,
  get: function () {
    return _objectUtils.fillNulls;
  }
});

var _guid = require('./lib/guid');

Object.defineProperty(exports, 'ensureGuid', {
  enumerable: true,
  get: function () {
    return _guid.ensureGuid;
  }
});
Object.defineProperty(exports, 'initializeGuid', {
  enumerable: true,
  get: function () {
    return _guid.initializeGuid;
  }
});

var _collections = require('./lib/collections');

Object.defineProperty(exports, 'Stack', {
  enumerable: true,
  get: function () {
    return _collections.Stack;
  }
});
Object.defineProperty(exports, 'DictSet', {
  enumerable: true,
  get: function () {
    return _collections.DictSet;
  }
});
Object.defineProperty(exports, 'dict', {
  enumerable: true,
  get: function () {
    return _collections.dict;
  }
});

var _listUtils = require('./lib/list-utils');

Object.defineProperty(exports, 'EMPTY_SLICE', {
  enumerable: true,
  get: function () {
    return _listUtils.EMPTY_SLICE;
  }
});
Object.defineProperty(exports, 'LinkedList', {
  enumerable: true,
  get: function () {
    return _listUtils.LinkedList;
  }
});
Object.defineProperty(exports, 'ListNode', {
  enumerable: true,
  get: function () {
    return _listUtils.ListNode;
  }
});
Object.defineProperty(exports, 'ListSlice', {
  enumerable: true,
  get: function () {
    return _listUtils.ListSlice;
  }
});

var _arrayUtils = require('./lib/array-utils');

Object.defineProperty(exports, 'A', {
  enumerable: true,
  get: function () {
    return _interopRequireDefault(_arrayUtils).default;
  }
});
Object.defineProperty(exports, 'EMPTY_ARRAY', {
  enumerable: true,
  get: function () {
    return _arrayUtils.EMPTY_ARRAY;
  }
});

var _weakmap = require('./lib/weakmap');

Object.defineProperty(exports, 'HAS_NATIVE_WEAKMAP', {
  enumerable: true,
  get: function () {
    return _weakmap.HAS_NATIVE_WEAKMAP;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
